PEP 752: Address feedback, round 5 (#4018)
python/peps: peps/pep-0752.rst

and the other is malicious on ``Y`` then the user would be unable to satisfy
their requirements without encountering a malicious package.

.. _digital-attestations:

Exclusive Reliance on Digital Attestations
------------------------------------------

The idea [5]_ here would be to solely rely on :pep:`740` attestations to verify
certain properties of dependencies, such as:

* The package was uploaded by a specific organization or user name
* The package was uploaded by an owner of a specific domain name
* The package was uploaded by a user with a specific email address

A general downside to this approach is that implementing this on custom package
repositories would require a significant amount of work. In the case of PyPI,
the prerequisite work on `Trusted Publishing`__ and then the
`PEP 740 implementation`__ itself took the equivalent of a full-time engineer
one year whose time was paid for by a corporate sponsor. This is unlikely to be

> implemented by enterprise because it's possible to implement reproducible

Reviewer: "...by enterprise..." reads a little awkward.  Did you mean *an* enterprise (i.e. a corporate organization)?

Author: Fixed, thanks!

builds already using simpler mechanisms, and such attestations are not very

> useful when everything is internally managed.

Author: Thanks!

__ https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/#acknowledgements
__ https://blog.trailofbits.com/2024/10/01/securing-the-software-supply-chain-with-the-slsa-framework/

> The amount of effort wouldn't matter for community projects because they
> wouldn't be able to maintain any infrastructure themselves in most cases.

Reviewer 2: ```suggestion
Community projects are unlikely to undertake this effort because they
would likely lack the resources to maintain the necessary infrastructure themselves.
```

Author: Thanks!

Another weakness is that a user attempting to perform a simple ``pip install``
would already be vulnerable to malicious packages. In order to overcome this,

> there would have to be the concept of implicit constraints, all of which would

Reviewer: I'm not sure I understand what "implicit constraints" means.

Reviewer 2: @ofek This was confusing to me as well.

Author: Fixed, thanks! I was originally just copying the terminology that Alyssa put forth.

impose certain UX or resolver logic upon every tool.

For example, package managers could be changed such that the first time a
package is installed the user would receive a confirmation prompt displaying
the attestation details. This would be very confusing and noisy, especially for
new users, and would be a breaking UX change for existing users. Many methods
of installation wouldn't work for this scenario such as installing from a
requirements file where the user would potentially be getting hundreds of
prompts.

One solution to make this less disruptive for the user would be to manually
maintain a list of trustworthy details (organization/user names, domain names,
email addresses, etc.). This could be discoverable by packages providing
`entry points`__, which corporate environments could install by default. This
has the major downside of not providing automatic guarantees which would limit
the usefulness for the average user who is more likely to be affected.

__ https://packaging.python.org/en/latest/specifications/entry-points/

An idea to provide automatic protection would be to have a set of default

> packages that would be installed by default. Each repository could offer a

Reviewer: "...by default" - with the double "default" maybe the second one could be "...installed without further explicit confirmation." ?

Author: Fixed, thanks!

and the other is malicious on ``Y`` then the user would be unable to satisfy
their requirements without encountering a malicious package.

.. _provenance-assertions:

> Exclusive Reliance on Provenance Assertions

Reviewer 2: Love the term "Provenance Assertions". It evokes what I would expect as a user.

-------------------------------------------

The idea here [5]_ would be to design a general purpose way for clients to make
provenance assertions to verify certain properties of dependencies, each with
custom syntax. Some examples:

* The package was uploaded by a specific organization or user name e.g.
  ``pip install "azure-loganalytics from microsoft"``
* The package was uploaded by an owner of a specific domain name e.g.
  ``pip install "google-cloud-compute from cloud.google.com"``
* The package was uploaded by a user with a specific email address e.g.
  ``pip install "aws-cdk-lib from contact@amazon.com"``
* The package matching a namespace was uploaded by an authorized party (this
  PEP)

> A fundamental downside here is that it doesn't play well with multiple

Reviewer 2: ```suggestion
A fundamental downside is that it doesn't play well with multiple
```

repositories. For example, say a user wants the ``azure-loganalytics`` package
and wants to ensure it comes from the organization named ``microsoft``. If
Microsoft's organization name on PyPI is ``microsoft`` then a package manager
that defaults to PyPI could accept ``azure-loganalytics from microsoft``.
However, if multiple repositories are used for dependency resolution then the
user would have to specify the repository as part of the definition which is
unrealistic for reasons outlined in the dedicated section on
`asserting package owner names <asserting-package-owner-names_>`_.

Another general weakness with this approach is that a user attempting to
perform a simple ``pip install`` without special syntax, which is the most

> common scenario, would already be vulnerable to malicious packages. In order to
> overcome this there would have to be some default trust mechanism, which in all
> cases would impose certain UX or resolver logic upon every tool.

Reviewer 2: ```suggestion
common scenario, would already be vulnerable to malicious packages. To
overcome this, some default trust mechanism, which in all
cases would impose certain UX or resolver logic upon every tool, would be needed.
```

Author: What about the following?

> To overcome this, a default trust mechanism is required, which in all cases would impose certain UX or resolver logic upon every tool.

Reviewer 2: I like your suggestion @ofek.
